File: minimap-lab/web/follower.js
// Route following state, kept free of the DOM so it can be tested directly.
// The panel feeds it positions from the tracker and paths from /api/path; it
// answers with what the player should do next.
const TRANSITION_INSTRUCTIONS = {
  rope: 'Użyj liny',
  ladder: 'Wejdź po drabinie',
  stairs: 'Wejdź na schody',
  hole: 'Zejdź dziurą',
  shovel: 'Kop łopatą',
  walk: 'Przejdź na piętro',
};
const COMPASS = [['NW', 'N', 'NE'], ['W', '', 'E'], ['SW', 'S', 'SE']];

const chebyshev = (a, b) => Math.max(Math.abs(a.x - b.x), Math.abs(a.y - b.y));

class RouteFollower {
  constructor(waypoints = [], options = {}) {
    this.waypoints = waypoints;
    this.tolerance = options.tolerance ?? 1;
    // An action waypoint needs the exact tile: a rope used one tile off the
    // rope spot does nothing, while walking tolerance may stay loose.
    this.actionTolerance = options.actionTolerance ?? 0;
    this.loop = options.loop ?? false;
    this.replanMS = options.replanMS ?? 500;
    // A failure is a moment, not a verdict: one dropped request must not freeze
    // guidance until the user toggles following off and on again.
    this.retryMS = options.retryMS ?? 4000;
    this.reset();
  }
  reset() {
    this.index = 0;
    this.finished = false;
    this.path = null;
    this.pathTo = null;
    this.blocked = null;
    this.requestedAt = null;
    this.actionAt = null;
  }
  get waypoint() {
    return this.finished ? null : this.waypoints[this.index] ?? null;
  }
  // skipTo jumps to a waypoint, dropping work done for the previous one.
  skipTo(index) {
    this.index = Math.max(0, Math.min(this.waypoints.length - 1, index));
    this.finished = false;
    this.actionAt = null;
    this.dropPath();
  }
  dropPath() {
    this.path = null;
    this.pathTo = null;
    this.blocked = null;
    this.blockedAt = null;
    this.requestedAt = null;
  }
  // askedFor is the waypoint the request was made for; a reply that took long
  // enough for the target to move describes a route the follower no longer wants.
  setPath(result, now, askedFor) {
    const target = this.waypoint;
    // Leave state untouched: whatever replaced this request is more current.
    if (!target || !askedFor || !sameTile(askedFor, target)) return;
    this.requestedAt = now;
    if (!result || !result.found) {
      this.path = null;
      this.blocked = result ? {status: result.status, reason: result.reason} : null;
      this.blockedAt = {at: now};
      return;
    }
    this.blocked = null;
    this.path = result.steps.map(s => [s[0], s[1]]);
    this.pathTo = this.waypoint && {...this.waypoint};
  }
  step(position, now) {
    const target = this.advance(position);
    if (!target) return {action: 'done'};
    // Two ways to end up waiting for a floor change: standing on a waypoint
    // that carries an action, or facing a waypoint on another floor.
    const standingOnAction = target.type !== 'walk' && target.z === position.z &&
      chebyshev(target, position) <= this.actionTolerance;
    if (standingOnAction || target.z !== position.z) {
      this.dropPath();
      if (standingOnAction) this.actionAt = {index: this.index, z: position.z};
      const verb = TRANSITION_INSTRUCTIONS[target.type] ?? TRANSITION_INSTRUCTIONS.walk;
      const floor = standingOnAction ? '' : ` → piętro ${target.z}`;
      return {action: 'transition', waypoint: target, next: this.waypoints[this.index + 1] ?? null,
        instruction: `${verb}${floor}`};
    }
    if (this.pathTo && !sameTile(this.pathTo, target)) this.dropPath();
    const ahead = this.path && remainingPath(this.path, position);
    if (ahead && ahead.length > 1) {
      return {
        action: 'walk',
        direction: direction(position, ahead[1]),
        next: ahead[1],
        remaining: ahead.length - 1,
        waypoint: target,
      };
    }
    if (this.blocked) {
      const moved = this.blockedAt?.from && !sameTile(this.blockedAt.from, position);
      const waited = now - this.blockedAt?.at >= this.retryMS;
      if (!moved && !waited) {
        this.blockedAt.from ??= {...position};
        return {action: 'blocked', waypoint: target, ...this.blocked};
      }
      this.blocked = null;
      this.blockedAt = null;
      this.requestedAt = null; // the situation changed, so ask again now
    }
    if (this.requestedAt !== null && now - this.requestedAt < this.replanMS) return {action: 'wait', waypoint: target};
    this.requestedAt = now;
    this.path = null;
    return {action: 'path', from: {...position}, to: {x: target.x, y: target.y, z: target.z}, waypoint: target};
  }
  // advance consumes every waypoint already reached and returns the next one.
  advance(position) {
    // One pass at most: a looped route whose points all sit within tolerance
    // would otherwise cycle forever and re-request a path on every reading.
    for (let visited = 0; !this.finished; visited++) {
      if (visited > this.waypoints.length) {
        this.finished = true;
        return null;
      }
      const target = this.waypoints[this.index];
      if (!target) {
        this.finished = true;
        return null;
      }
      // An action waypoint is done once the action has been carried out, which
      // the tracker reports as a change of floor. The player can also cross
      // between two readings, never being seen standing on the waypoint - then
      // standing on the floor the route continues on is the evidence.
      const armed = this.actionAt?.index === this.index && this.actionAt.z !== position.z;
      const next = this.waypoints[this.index + 1];
      const crossed = target.type !== 'walk' && target.z !== position.z && next?.z === position.z;
      const acted = armed || crossed;
      if (acted) this.actionAt = null;
      if (!acted) {
        if (target.z !== position.z || chebyshev(target, position) > this.tolerance) return target;
        if (target.type !== 'walk') return target;
      }
      this.dropPath();
      if (this.index + 1 < this.waypoints.length) {
        this.index++;
      } else if (this.loop && this.waypoints.length) {
        this.index = 0;
      } else {
        this.finished = true;
        return null;
      }
    }
    return null;
  }
}

const sameTile = (a, b) => a.x === b.x && a.y === b.y && a.z === b.z;

// remainingPath cuts the path back to the player's tile. A player standing off
// the path gets null, which sends the follower back to /api/path.
function remainingPath(path, position) {
  const at = path.findIndex(s => s[0] === position.x && s[1] === position.y);
  return at < 0 ? null : path.slice(at);
}

function direction(from, next) {
  const dx = Math.sign(next[0] - from.x) + 1;
  const dy = Math.sign(next[1] - from.y) + 1;
  return COMPASS[dy][dx];
}

globalThis.RouteFollower = RouteFollower;
if (typeof module !== 'undefined') module.exports = {RouteFollower};

File: minimap-lab/internal/brain/follower.go
package brain

import (
	"fmt"
	"time"

	"minimap-lab/internal/mapdata"
	"minimap-lab/internal/nav"
	"minimap-lab/internal/route"
)

var transitionInstructions = map[string]string{
	"rope":   "Użyj liny",
	"ladder": "Wejdź po drabinie",
	"stairs": "Wejdź na schody",
	"hole":   "Zejdź dziurą",
	"shovel": "Kop łopatą",
	"walk":   "Przejdź na piętro",
}

func chebyshev(w route.Waypoint, p mapdata.Position) int {
	dx, dy := abs(w.X-p.X), abs(w.Y-p.Y)
	if dx > dy {
		return dx
	}
	return dy
}

func sameTile(a, b route.Waypoint) bool { return a.X == b.X && a.Y == b.Y && a.Z == b.Z }

type FollowerOptions struct {
	// Tolerance is a pointer for the same reason pathapi.go makes coordinates
	// pointers: an unset field has to be distinguishable from a deliberate
	// zero. Nil means the default of one tile; zero means the character must
	// land exactly on the waypoint, which is a legitimate strict setting.
	Tolerance *int
	// ActionTolerance needs no such treatment - its default is zero anyway. A
	// rope used one tile off the rope spot does nothing at all, while walking
	// tolerance may stay loose.
	ActionTolerance int
	Loop            bool
	Replan          time.Duration
	Retry           time.Duration
}

type blockedInfo struct {
	status, reason string
}

// Follower turns a route and a position into what the player should do next.
// It knows nothing about keyboards or timeouts - only about where the route
// goes and whether a usable path to the next waypoint exists.
type Follower struct {
	waypoints []route.Waypoint
	tolerance int
	// actionTolerance is read on its own rather than falling back to
	// tolerance: the two answer different questions.
	actionTolerance int
	loop            bool
	replan          time.Duration
	// retry exists because a failure is a moment, not a verdict: one dropped
	// request must not freeze guidance until the user toggles following off
	// and on again.
	retry time.Duration

	index    int
	finished bool

	path   [][2]int
	pathTo *route.Waypoint

	blocked      *blockedInfo
	blockedAt    time.Time
	blockedFrom  *mapdata.Position
	hasBlockedAt bool

	requestedAt    time.Time
	hasRequestedAt bool

	// actionAt remembers that the character was standing on an action waypoint
	// and on which floor, so the floor change that follows can be recognised as
	// that action having been carried out.
	actionAt      *actionArmed
	minOverlayRev uint64
}

type actionArmed struct {
	index int
	z     int
}

func NewFollower(waypoints []route.Waypoint, o FollowerOptions) *Follower {
	f := &Follower{
		waypoints:       waypoints,
		tolerance:       1,
		actionTolerance: o.ActionTolerance,
		loop:            o.Loop,
		replan:          o.Replan,
		retry:           o.Retry,
	}
	if o.Tolerance != nil {
		f.tolerance = *o.Tolerance
	}
	if f.replan == 0 {
		f.replan = 500 * time.Millisecond
	}
	if f.retry == 0 {
		f.retry = 4 * time.Second
	}
	return f
}

func (f *Follower) Index() int     { return f.index }
func (f *Follower) Finished() bool { return f.finished }
func (f *Follower) Path() [][2]int { return f.path }

func (f *Follower) SetMinOverlayRevision(rev uint64) { f.minOverlayRev = rev }

func (f *Follower) Waypoint() *route.Waypoint {
	if f.finished || f.index < 0 || f.index >= len(f.waypoints) {
		return nil
	}
	return &f.waypoints[f.index]
}

func (f *Follower) Reset() {
	f.index, f.finished = 0, false
	f.DropPath()
	f.actionAt = nil
}

// SkipTo jumps to a waypoint, dropping work done for the previous one.
func (f *Follower) SkipTo(i int) {
	if len(f.waypoints) == 0 {
		return
	}
	if i < 0 {
		i = 0
	}
	if i > len(f.waypoints)-1 {
		i = len(f.waypoints) - 1
	}
	f.index, f.finished = i, false
	// A skipped-away action must be armed again rather than remembered: the
	// floor change that follows belongs to whatever the route is doing now.
	f.actionAt = nil
	f.DropPath()
}

func (f *Follower) DropPath() {
	f.path, f.pathTo = nil, nil
	f.blocked, f.blockedFrom, f.hasBlockedAt = nil, nil, false
	f.hasRequestedAt = false
}

// SetPath installs a planner reply. askedFor is the waypoint the request was
// made for; a reply that took long enough for the target to move describes a
// route the follower no longer wants.
func (f *Follower) SetPath(res *nav.PathResult, now time.Time, askedFor *route.Waypoint) {
	target := f.Waypoint()
	// Leave state untouched: whatever replaced this request is more current.
	if target == nil || askedFor == nil || !sameTile(*askedFor, *target) {
		return
	}
	// Same reasoning, one dimension further: the waypoint may be unchanged
	// while the map underneath it is not. Revision zero means the reply
	// carried none - a locally produced failure - and is also the state of a
	// store that has learned nothing, so nothing can be stale against it.
	if res != nil && res.OverlayRevision > 0 && res.OverlayRevision < f.minOverlayRev {
		return
	}
	f.requestedAt, f.hasRequestedAt = now, true
	if res == nil || !res.Found {
		f.path, f.pathTo = nil, nil
		f.blocked = nil
		if res != nil {
			f.blocked = &blockedInfo{status: res.Status, reason: res.Reason}
		}
		f.blockedAt, f.hasBlockedAt, f.blockedFrom = now, true, nil
		return
	}
	f.blocked = nil
	f.path = make([][2]int, len(res.Steps))
	copy(f.path, res.Steps)
	kept := *target
	f.pathTo = &kept
}

func (f *Follower) Step(p mapdata.Position, now time.Time) Output {
	target := f.advance(p)
	if target == nil {
		return Output{Action: ActionDone}
	}
	standingOnAction := target.Type != "walk" && target.Z == p.Z && chebyshev(*target, p) <= f.actionTolerance
	if standingOnAction || target.Z != p.Z {
		f.DropPath()
		if standingOnAction {
			f.actionAt = &actionArmed{index: f.index, z: p.Z}
		}
		verb, ok := transitionInstructions[target.Type]
		if !ok {
			verb = transitionInstructions["walk"]
		}
		floor := ""
		if !standingOnAction {
			floor = fmt.Sprintf(" → piętro %d", target.Z)
		}
		out := Output{Action: ActionTransition, Index: f.index, Waypoint: target,
			Instruction: verb + floor}
		if f.index+1 < len(f.waypoints) {
			out.NextWaypoint = &f.waypoints[f.index+1]
		}
		return out
	}
	if f.pathTo != nil && !sameTile(*f.pathTo, *target) {
		f.DropPath()
	}
	if ahead := remainingPath(f.path, p); len(ahead) > 1 {
		return Output{Action: ActionWalk, Direction: directionTo(p, ahead[1]),
			Next: ahead[1], Remaining: len(ahead) - 1, Waypoint: target}
	}
	if f.blocked != nil {
		moved := f.blockedFrom != nil && *f.blockedFrom != p
		waited := f.hasBlockedAt && now.Sub(f.blockedAt) >= f.retry
		if !moved && !waited {
			if f.blockedFrom == nil {
				here := p
				f.blockedFrom = &here
			}
			return Output{Action: ActionBlocked, Waypoint: target,
				Status: f.blocked.status, Reason: f.blocked.reason}
		}
		// The situation changed, so ask again now rather than waiting out the
		// replan throttle on top of the backoff already served.
		f.blocked, f.blockedFrom, f.hasBlockedAt = nil, nil, false
		f.hasRequestedAt = false
	}
	if f.hasRequestedAt && now.Sub(f.requestedAt) < f.replan {
		return Output{Action: ActionWait, Waypoint: target}
	}
	f.requestedAt, f.hasRequestedAt = now, true
	f.path = nil
	return Output{Action: ActionPath, From: p,
		To: mapdata.Position{X: target.X, Y: target.Y, Z: target.Z}, Waypoint: target}
}

// advance consumes every waypoint already reached and returns the next one.
func (f *Follower) advance(p mapdata.Position) *route.Waypoint {
	// One pass at most: a looped route whose points all sit within tolerance
	// would otherwise cycle forever and re-request a path on every reading.
	for visited := 0; !f.finished; visited++ {
		if visited > len(f.waypoints) {
			f.finished = true
			return nil
		}
		if f.index < 0 || f.index >= len(f.waypoints) {
			f.finished = true
			return nil
		}
		target := &f.waypoints[f.index]
		// An action waypoint is done once the action has been carried out,
		// which the tracker reports as a change of floor. The player can also
		// cross between two readings, never being seen standing on the
		// waypoint - then standing on the floor the route continues on is the
		// evidence.
		armed := f.actionAt != nil && f.actionAt.index == f.index && f.actionAt.z != p.Z
		crossed := false
		if target.Type != "walk" && target.Z != p.Z && f.index+1 < len(f.waypoints) {
			crossed = f.waypoints[f.index+1].Z == p.Z
		}
		acted := armed || crossed
		if acted {
			f.actionAt = nil
		}
		if !acted {
			if target.Z != p.Z || chebyshev(*target, p) > f.tolerance {
				return target
			}
			if target.Type != "walk" {
				return target
			}
		}
		f.DropPath()
		switch {
		case f.index+1 < len(f.waypoints):
			f.index++
		case f.loop && len(f.waypoints) > 0:
			f.index = 0
		default:
			f.finished = true
			return nil
		}
	}
	return nil
}

// remainingPath cuts the path back to the player's tile. A player standing off
// the path gets nothing, which sends the follower back to the planner.
func remainingPath(path [][2]int, p mapdata.Position) [][2]int {
	for i, s := range path {
		if s[0] == p.X && s[1] == p.Y {
			return path[i:]
		}
	}
	return nil
}

func directionTo(from mapdata.Position, next [2]int) string {
	return executorCompass[sign(next[1]-from.Y)+1][sign(next[0]-from.X)+1]
}

// SetOptions adjusts the follower in place. Rebuilding it would be simpler but
// would throw away which waypoint the route had reached, so a user nudging the
// tolerance mid-route would be sent back to the start.
func (f *Follower) SetOptions(o FollowerOptions) {
	if o.Tolerance != nil {
		f.tolerance = *o.Tolerance
	}
	f.actionTolerance = o.ActionTolerance
	f.loop = o.Loop
	if o.Replan > 0 {
		f.replan = o.Replan
	}
	if o.Retry > 0 {
		f.retry = o.Retry
	}
}
